src/workerd/api/encoding.c++

```diff
@@ -460,29 +461,204 @@ kj::Maybe<jsg::JsString> TextDecoder::decodePtr(
 // =======================================================================================
 // TextEncoder implementation
 
-jsg::Ref<TextEncoder> TextEncoder::constructor(jsg::Lock& js) {
-  return js.alloc<TextEncoder>();
+namespace {
+
+constexpr inline bool isLeadSurrogate(char16_t c) {
+  return 0xD800 <= c && c < 0xDC00;
 }
 
-namespace {
-TextEncoder::EncodeIntoResult encodeIntoImpl(
-    jsg::Lock& js, jsg::JsString input, jsg::BufferSource& buffer) {
-  auto result = input.writeInto(
-      js, buffer.asArrayPtr().asChars(), jsg::JsString::WriteFlags::REPLACE_INVALID_UTF8);
-  return TextEncoder::EncodeIntoResult{
-    .read = static_cast<int>(result.read),
-    .written = static_cast<int>(result.written),
-  };
+constexpr inline bool isTrailSurrogate(char16_t c) {
+  return 0xDC00 <= c && c <= 0xDFFF;
+}
+
+// Calculate the number of UTF-8 bytes needed for a single UTF-16 code unit
+constexpr inline size_t utf8BytesForCodeUnit(char16_t c) {
+  if (c < 0x80) return 1;
+  if (c < 0x800) return 2;
+  return 3;
 }
+
+// Calculate UTF-8 length from UTF-16 with potentially invalid surrogates.
+// Invalid surrogates are counted as U+FFFD (3 bytes in UTF-8).
+size_t utf8LengthFromInvalidUtf16(const char16_t* input, size_t length) {
+  size_t utf8Length = 0;
+  bool pendingSurrogate = false;
+
+  for (size_t i = 0; i < length; i++) {
+    char16_t c = input[i];
+
+    if (pendingSurrogate) {
+      if (isTrailSurrogate(c)) {
+        // Valid surrogate pair = 4 bytes in UTF-8
+        utf8Length += 4;
+        pendingSurrogate = false;
+      } else {
+        // Unpaired lead surrogate = U+FFFD (3 bytes)
+        utf8Length += 3;
+        if (!isLeadSurrogate(c)) {
+          utf8Length += utf8BytesForCodeUnit(c);
+          pendingSurrogate = false;
+        }
+      }
+    } else if (isLeadSurrogate(c)) {
+      pendingSurrogate = true;
+    } else {
+      if (isTrailSurrogate(c)) {
+        // Unpaired trail surrogate = U+FFFD (3 bytes)
+        utf8Length += 3;
+      } else {
+        utf8Length += utf8BytesForCodeUnit(c);
+      }
+    }
+  }
+
+  if (pendingSurrogate) {
+    utf8Length += 3;  // Trailing unpaired lead surrogate
+  }
+
+  return utf8Length;
+}
+
+// Encode a single UTF-16 code unit to UTF-8
+inline size_t encodeUtf8CodeUnit(char16_t c, char* out) {
+  if (c < 0x80) {
+    *out = static_cast<char>(c);
+    return 1;
+  } else if (c < 0x800) {
+    out[0] = static_cast<char>(0xC0 | (c >> 6));
+    out[1] = static_cast<char>(0x80 | (c & 0x3F));
+    return 2;
+  } else {
+    out[0] = static_cast<char>(0xE0 | (c >> 12));
+    out[1] = static_cast<char>(0x80 | ((c >> 6) & 0x3F));
+    out[2] = static_cast<char>(0x80 | (c & 0x3F));
+    return 3;
+  }
+}
+
+// Encode a valid surrogate pair to UTF-8
+inline void encodeSurrogatePair(char16_t lead, char16_t trail, char* out) {
```

Comment:
Ditto here... `kj::ArrayPtr` instead of `char*`. Will stop pointing this out... comments applies in each of the places a raw pointer is being passed in.